### PR TITLE
feat: add config db-hoisted-tx-settings to allow only hoisted function settings

### LINKED ISSUE #3242: Regression / breaking change for RPCs with SET
### Environment

* PostgreSQL version: 15.x
* PostgREST version: 12.1-dev
* Operating system: Linux

### Description of issue

#3142 introduced quite a big breaking change - which I would consider a regression / bug. Applying **all** settings is never going to be correct / expected.

Take this for example:

```sql
create function do_something_dangerous() returns something
security definer
set search_path to pg_catalog, pg_temp
language plpgsql as $$
...
$$;
```

I have a few `SECURITY DEFINER` functions in my api. Best practice for security reasons is to add `pg_catalog, pg_temp` as a search_path for those kind of functions. After #3142, this suddenly changes the search_path for the **whole transaction**. This breaks, for example, computed fields on the returned value - and probably more.

I do have other cases of `SET`tings that I **don't want to have applied on the transaction level**. There is a reason I set them on the function and not for the authenticator role or so.

While I like the idea of hoisting some `SET`tings to the transaction level.. the current implementation is a problem. I don't even see a check for `role`, so will a `CREATE FUNCTION xxx SET ROLE TO y` be hoisted, too? Bad.

I think we should do at least one of, but possibly both of:
- Have an allowlist of explicitly named parameters instead of just setting all of them.
- Prefix those settings with `pgrst.xxx` or `tx.xxx` - and then set it without that prefix on the transaction level.

This needs to happen before we make the next release, otherwise we can't release a minor. In the current form, this is quite a breaking change.

### PR DESCRIPTION
**Pull request by taimoorzaeem**
Fixes #3242.

@steve-chavez @laurenceisla @wolfgangwalther Even though I have applied the filter for function settings using the new config, but the setting is still being applied in `io tests`. Am I testing it right?

### CODE REVIEW

File: test/io/fixtures.sql
Comment (taimoorzaeem) [reply test/io/fixtures.sql:222]: @wolfgangwalther How about this?
```sql
-- returns computed fields
create or replace function not_hoisted_should_not_apply()
returns table (
  work_mem text,
  statement_timeout text
)
as $$
begin
  return query
          select
            current_setting('work_mem',false) as work_mem,
            current_setting('statement_timeout',false) as statement_timeout;
end;
$$ language plpgsql
set work_mem = '5000'
set statement_timeout = '5s';

-- calls the above function
create or replace function test_not_hoisted_should_not_apply()
returns text
as $$
  select work_mem || ' ' || statement_timeout
  from not_hoisted_should_not_apply();
$$ language sql
```
Calling `test_not_hoisted_should_not_apply` yields `5000kb 5s` which is strange because only `work_mem` is hoisted but `statement_timeout` is still being applied. Is this right way to write computed fields?

Comment (taimoorzaeem) [reply test/io/fixtures.sql:222]: @wolfgangwalther Thanks a lot! It worked.

**Comment by taimoorzaeem** [comment]
It appears that the test case is failing, however, the test passes when run with `postgrest-test-io -k test_only_hoisted_settings_are_applied`.

**Comment by taimoorzaeem** [comment]
> It seems like you need to update the pytest snapshot for schema cache, too.

I am not sure what you mean by that. How do I update pytest?

**Comment by taimoorzaeem** [comment]
> Or maybe some other test sets the statement timeout to `5s`? What if you change the timeout on the RPC to e.g. `6s`? Does wrong expectation change as well? Then you'd be sure that this is actually the value we are talking about.

Thanks for this suggestion. I tried the changing the value to `6s` and `7s` and it still showed `5s` at test time. My guess is that `statement_timeout` has default values of `5s` when all `io-tests` are run and `2s` when an individual io-test is run with `-k` flag. I've changed the value to `5s` to indicate default value in the test.

File: src/PostgREST/Query.hs
Comment (taimoorzaeem) [reply src/PostgREST/Query.hs]: @wolfgangwalther By the changing the filter, io-tests are failing as the config is not applying now.
```sql
LEFT JOIN LATERAL (
    SELECT
      array_agg(row(
        substr(setting, 1, strpos(setting, '=') - 1),
        lower(substr(setting, strpos(setting, '=') + 1))
      )) as kvs
    FROM unnest(proconfig) setting
    WHERE setting NOT LIKE ALL(ARRAY['|] <> encodeUtf8 (T.intercalate "," hoistedSets) <> [q|'])
  ) func_settings ON TRUE
  WHERE t.oid <> 'trigger'::regtype AND COALESCE(a.callable, true)
|] <> (if pgVer >= pgVersion110 then "AND prokind = 'f'" else "AND NOT (proisagg OR proiswindow)")
```
I get,
```
FAILED test/io/test_io.py::test_first_hoisted_setting_is_applied - assert '{"get_work_m...imeout":"7s"}' == '{"get_work_m...imeout":"2s"}'
FAILED test/io/test_io.py::test_second_hoisted_setting_is_applied - assert '{"get_work_m...imeout":"7s"}' == '{"get_work_m...imeout":"7s"}'
============= 2 failed, 157 passed, 1 skipped in 67.99s (0:01:07) ==============

```

Comment (laurenceisla) [reply src/PostgREST/Query.hs]: It's not applying because:

```sql
    WHERE setting NOT LIKE ALL(ARRAY['|] <> encodeUtf8 (T.intercalate "," hoistedSets) <> [q|'])
```

Will return:

```sql
    WHERE setting NOT LIKE ALL(ARRAY['statement_timeout,plan_filter.statement_cost_limit,default_transaction_isolation'])
```

That is, all the settings as a single element of the array.

Also, since this is a dynamic value, I think this should be parameterized. You can do something similar to what it's done here:

https://github.com/PostgREST/postgrest/blob/11c9e8dac903aa27511defa390383bcecd256681/src/PostgREST/SchemaCache.hs#L937-L939
https://github.com/PostgREST/postgrest/blob/11c9e8dac903aa27511defa390383bcecd256681/src/PostgREST/SchemaCache.hs#L980

**Comment by steve-chavez** [comment]
To unblock v12.2.0 (https://github.com/PostgREST/postgrest/issues/3501), just rebased and addressed the latest reviews.

File: docs/references/configuration.rst
Comment (laurenceisla) [inline docs/references/configuration.rst:311]: Here it mentions that it can be set by in-db config, but it's not added to:

https://github.com/PostgREST/postgrest/blob/5d3d09923f405bf53c6384a4487fee9195c3118f/src/PostgREST/Config/Database.hs#L44-L46

And there are no entries in  `db_config.sql`  to test this.

Comment (steve-chavez) [reply docs/references/configuration.rst:311]: Done

File: test/io/db_config.sql
Comment (laurenceisla) [inline test/io/db_config.sql:26]: This should be the last one before merging. The comment mentions:

> -- these settings will override the values in configs/no-defaults.config, so they must be different

So this should be different from the value on `no-defaults.config` to test that it's actually changing.

Comment (steve-chavez) [reply test/io/db_config.sql:26]: Ok done.

Btw this whole workflow is double plus ungood, adding a config option is always painful.

This and https://github.com/PostgREST/postgrest/pull/3481#issuecomment-2123572322 tells us the whole Config handling needs reworking.

**Review by laurenceisla — APPROVED** [review]

**Pull request merged**
